openmeter/billing/invoicelinesplitgroup.go

```diff
@@ -0,0 +1,282 @@
+}
+
+func (i SplitLineGroup) Clone() SplitLineGroup {
+	return SplitLineGroup{
+		ManagedModel:       i.ManagedModel,
+		SplitLineGroupBase: i.SplitLineGroupBase.Clone(),
+		ID:                 i.ID,
+	}
+}
+
+type LineWithInvoiceHeader struct {
+	Line    *Line
+	Invoice InvoiceBase
+}
+
+func (i LineWithInvoiceHeader) Clone() LineWithInvoiceHeader {
+	return LineWithInvoiceHeader{
+		Line:    i.Line.Clone(),
+		Invoice: i.Invoice,
+	}
+}
+
+type SplitLineHierarchy struct {
+	Group SplitLineGroup
+	Lines []LineWithInvoiceHeader
+}
+
+func (h *SplitLineHierarchy) Clone() SplitLineHierarchy {
+	return SplitLineHierarchy{
+		Group: h.Group.Clone(),
+		Lines: lo.Map(h.Lines, func(line LineWithInvoiceHeader, _ int) LineWithInvoiceHeader {
+			return line.Clone()
+		}),
+	}
+}
+
+type SumNetAmountInput struct {
+	PeriodEndLTE   time.Time
+	IncludeCharges bool
+}
+
+// SumNetAmount returns the sum of the net amount (pre-tax) of the progressive billed line and its children
+// containing the values for all lines whose period's end is <= in.UpTo and are not deleted or not part of
+// an invoice that has been deleted.
+func (h *SplitLineHierarchy) SumNetAmount(in SumNetAmountInput) alpacadecimal.Decimal {
+	netAmount := alpacadecimal.Zero
+
+	_ = h.ForEachChild(ForEachChildInput{
+		PeriodEndLTE: in.PeriodEndLTE,
+		Callback: func(child LineWithInvoiceHeader) error {
+			netAmount = netAmount.Add(child.Line.Totals.Amount)
+
+			if in.IncludeCharges {
+				netAmount = netAmount.Add(child.Line.Totals.ChargesTotal)
+			}
+
+			return nil
+		},
+	})
+
+	return netAmount
+}
+
+type ForEachChildInput struct {
+	PeriodEndLTE time.Time
+	Callback     func(child LineWithInvoiceHeader) error
+}
+
+func (h *SplitLineHierarchy) ForEachChild(in ForEachChildInput) error {
+	for _, child := range h.Lines {
+		// The line is not in scope
+		if !in.PeriodEndLTE.IsZero() && child.Line.Period.End.After(in.PeriodEndLTE) {
+			continue
+		}
+
+		// The line is deleted
+		if child.Line.DeletedAt != nil {
+			continue
+		}
+
+		// The invoice is deleted
+		if child.Invoice.DeletedAt != nil || child.Invoice.Status == InvoiceStatusDeleted {
+			continue
+		}
+
+		if err := in.Callback(child); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Adapter
+type (
+	CreateSplitLineGroupAdapterInput = SplitLineGroupBase
+	UpdateSplitLineGroupInput        SplitLineGroup
+	DeleteSplitLineGroupInput        = models.NamespacedID
+	GetSplitLineGroupInput           = models.NamespacedID
+)
+
+func (i UpdateSplitLineGroupInput) Validate() error {
+	err := i.SplitLineGroupBase.Validate()
+	if err != nil {
+		return err
+	}
+
+	if i.ID == "" {
+		return errors.New("id is required")
+	}
+
+	return nil
+}
+
+// LineOrHierarchy is a wrapper around a line or a split line hierarchy
+
+type LineOrHierarchyType string
+
+const (
+	LineOrHierarchyTypeLine      LineOrHierarchyType = "line"
+	LineOrHierarchyTypeHierarchy LineOrHierarchyType = "hierarchy"
+)
+
+type LineOrHierarchy struct {
```

Comment:
I'm slightly confused why the public API is the struct and the interface the internal and not the other way around.

Reply:
it could be:

```
type Line interface {
  Type()
  AsCharge()
  AsGroup()
}
```